python/ray/serve/deployment_state.py

```diff
@@ -1098,17 +1097,22 @@ def deploy(self, deployment_info: DeploymentInfo) -> bool:
                 and existing_info.version == deployment_info.version
             ):
                 return False
+        # Save the current status so we can revert in case of failure
+        orig = {k: v for (k, v) in self.__dict__.items()}
+        self._set_deployment_goal(deployment_info)
+        try:
+            # NOTE(edoakes): we must write a checkpoint before starting new
+            # or pushing the updated config to avoid inconsistent state if we
+            # crash while making the change.
+            self._save_checkpoint_func()
+        except Exception:
```

Comment:
@edoakes We've removed rollback behavior in the past since we wanted to handle at a higher level via the REST API. What do you think about performing a rollback here when writing to the GCS fails?

Reply:
do we need to rollback? `_save_checkpoint_func` should not be state mutating

Reply:
@simon-mo we need. The thing here is that we update the memory, but not checkpointing so the data in memory won't match the one in the storage. The state mutation is about the memory data.

In this case, if we redeploy with version=2, the new code won't deploy.

Reply:
Btw, I believe for all checkpointing in serve, we should revisit whether the rollback is needed. I'm not very familiar with that, so I'll leave the work to serve team.

Reply:
The simplest solution I can think of here is to write ahead to the GCS instead of first mutating internal state. That way, changes will only go through once they've been persisted. If we crash after GCS write, we'll recover the updated state anyways.

Reply:
@edoakes  That makes sense, but it seems that the checkpointing code is not doing it in that way.  What I'm doing here is similar to what you mentioned. Do you have some suggestion about how to do this? I hope serve can help on this.

Reply:
does it mean we just need to reorder the `self._save_checkpoint_func()` and `self._set_deployment_goal(deployment_info)`?

Reply:
@scv119 it's not quite that simple, because we currently checkpoint the in-memory state, which is set by `set_deployment_goal`...

Reply:
we'll need to refactor such that we have some explicit goal state object that we checkpoint, then set.